core/src/main/java/io/questdb/griffin/DatabaseSnapshotAgent.java

```diff
@@ -269,8 +268,22 @@ public void prepareSnapshot(SqlExecutionContext executionContext) throws SqlExce
                 try (RecordCursor cursor = factory.getCursor(executionContext)) {
                     final Record record = cursor.getRecord();
                     try (MemoryCMARW mem = Vm.getCMARWInstance()) {
-                        // Copy metadata files for all tables.
 
+                        // Suspend the WalPurgeJob
+                        if (walPurgeJobRunLock != null) {
+                            final long timeout = configuration.getCircuitBreakerConfiguration().getTimeout();
+                            try {
+                                if (!walPurgeJobRunLock.tryLock(timeout, TimeUnit.MICROSECONDS)) {
+                                    LOG.error().$("snapshot failed to acquire lock").$();
+                                    throw SqlException.position(0).put("Snapshot command cancelled due to timeout");
+                                }
+                            } catch (InterruptedException e) {
+                                LOG.error().$("snapshot error [e=").$(e).I$();
+                                throw SqlException.position(0).put("Snapshot command was interrupted");
+                            }
+                        }
+
+                        // Copy metadata files for all tables.
```

Comment:
Don't we need to copy WAL metadata as well? I mean sequencer metadata and table transaction log. Otherwise, we have no control over which blocks are flushed to disk after we have called `sync()`, so the actual fs snapshot may contain corrupted or missing data.

Reply:
Here I'm not very sure, but as far as I remember, we had a conversation with Alex on this topic, and we came to the conclusion that the metadata dump is enough.

Reply:
I cannot think of scenario where WAL transaction / metadata can have corrupt record. What do you have in mind @puzpuzpuz ? My perception that WAL transactions are atomic, we write record first and then make it visible.

Reply:
OS doesn't provide any order guarantees for the dirty page flushing. So, while we have the order (think, happens before) in-memory, the changes may end up on disk in any given order. That's why we're copying metadata files from in-memory to separate files for non-WAL tables and then call `flush()`. I don't think that WAL metadata and txn log files are an exception.

P.S. I've just realized that we may need to reject/block CREATE TABLE statements for non-WAL and WAL tables until SNAPSHOT COMPLETE. That's because those newly created tables aren't included in the snapshot, so we may get any bugs around them.

Reply:
What is the scenario you have in mind, can you add more details please?

Disk snapshot will include the dirty buffers won't it?

Reply:
> What is the scenario you have in mind, can you add more details please?

1. SNAPSHOT PREPARE executes
2. Users inserts data into a WAL table and commits it
3. The disk snapshot starts and "sees" the metadata changes, but only partially written latest row data
4. SNAPSHOT COMPLETE executes
5. Later, the user restores from the snapshot and sees some garbage instead of the latest rows

That's only one of the scenarios and there may be much more of them since the snapshot would be non-deterministic.

> Disk snapshot will include the dirty buffers won't it?

No, it only includes what's flushed to the "disk". The disk is a network-attached block storage, i.e. a separate machine that knows nothing of local dirty pages and OS page cache in general, and that's where the snapshot happens.

Reply:
Thanks.

I don't see metadata change in the scenario, only row insert. If you do WAL metadata change we perhaps have to do disk sync at the point when it is committed as a solution to sustain the reboots anyway.

Data commit is in WAL segments and WAL transaction log, if the WAL transaction log is in written but WAL segment is incomplete we will need a snapshot watermark. We should be able to rebuild WAL metadata from table meta + transaction log.

What do you suggest as a solution?

Reply:
> I don't see metadata change in the scenario

WAL txn log is also metadata and it certainly changes.

> What do you suggest as a solution?

As I mentioned in the first message, the solution might be as simple as copying all metadata files (sequencer metadata + txn logs), but alternatively we could copy sequencer metadata only and create a file with safe watermarks for each WAL txn log. Then we could use them when restoring for a snapshot.

Reply:
Txn log can be big file. And it's append only, so would be silly to copy it. Restore becomes non-trivial then.

Reply:
That's why I mentioned the watermarks.